### PR TITLE
add shape operator

### PR DESCRIPTION
**Pull request by mccanne**
This commit adds a Z operator to shape values of null:null into
their proper null type and coerces float64 to either uint64
or int64 if possible.  It does so by examining records that have
all the same field names where all numeric types are float64,
which is the case for unshaped csv and ndjson input.  It reads
the entire input, possibly spilling to disk, to determine the
proper types of everything then outputs the shaped values.
This is useful for cleaning up CSV inputs with various empty
values (i.e., null:null) appearing in various orders causes
the CSV reader to otherwise create many different record types
with varying null-typed columns.

We are generically calling this operator "shape" even though it's
functionality is initially limited as we expect this to be the seed 
for an operator that will do many interesting stateful shaping operations.

### CODE REVIEW

File: proc/shape/shaper.go [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+}
+
+func (a *anchor) updateInts(rec *zng.Record) error {
+	it := rec.Raw.Iter()
+	for k := range a.columns {
+		bytes, _, err := it.Next()
+		if err != nil {
+			return err
+		}
+		zv := zng.Value{rec.Type.Columns[k].Type, bytes}
```
Comment (nwt) [inline proc/shape/shaper.go:94]:
```suggestion
	for k, c := rec.Type.Columns {
		bytes, _, err := it.Next()
		if err != nil {
			return err
		}
		zv := zng.Value{c.Type, bytes}
```

File: proc/shape/shaper.go [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+	}
+	return nil
+}
+
+func (s *Shaper) newAnchor(columns []zng.Column) *anchor {
+	h := hash(&s.hash, columns)
+	a := &anchor{columns: columns}
+	a.next = s.anchors[h]
+	s.anchors[h] = a
+	a.integers = make([]integer, len(columns))
```
Comment (nwt) [inline proc/shape/shaper.go:165]:
```suggestion
	a := &anchor{
		columns:  columns,
		integers: make([]integer, len(columns)),
		next:     s.anchors[h],
	}
	s.anchors[h] = a
```

**Review by nwt — APPROVED** [review]

File: proc/shape/shape.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+	}
+	return nil, p.pctx.Err()
+}
+
+func (p *Proc) run() {
+	if err := p.pullInput(); err != nil {
+		p.shutdown(err)
+		return
+	}
+	p.shutdown(p.pushOutput())
```
[the remark this answers is not in the document]
Comment (mccanne) [reply proc/shape/shape.go:45]: I copied this pattern from elsewhere so we should fix it there too and/or do DRY

[the remark this answers is not in the document]
Comment (nwt) [reply proc/shape/shape.go:45]: It certainly does feel like there's some common code that we could factored out of the other, similar procs.

**Pull request merged**